doc/loginflow.md

```diff
@@ -0,0 +1,313 @@
+        "bakery-version": 3,
+        "cli-args": "[...]",
+        "user-data": "",
+        "client-version": "2.9.26"
+    }
+}
+```
+
+Because there are no credentials or macaroons in the request, the response
+contains a macaroon with a third party caveat that needs to be discharged to
+authenticate the user. This macaroon is made in
+`apiserver/authentication/user.go`, `UserAuthenticator.authenticateMacaroons()`.
+
+From looking at that code, The condition of the 3rd party caveat is
+`need-declared username admin`
+
+> Q: I guess it's encoded in the "v64" value below?
+
+Response:
+```json
+{
+    "request-id": 1,
+    "response": {
+        "discharge-required": {
+            "c": [
+                {
+                    "i": "time-before 2022-03-10T09:26:13.554951585Z"
+                },
+                {
+                    "i64": "AwA",
+                    "v64": "84lQ1w2X1y5_pzw6J43_UgGHkExE_T97jn95tTHEZAgaJW3IgRuNI7Yk0qpLM11_kPQ497PxW9yiv6sShpvgUYg7vgPRHi0r",
+                    "l": "https://CONTROLLER_HOST:17070/auth"
+                }
+            ],
+            "l": "juju model 10c91043-b22b-4e62-8f44-30132106b057",
+            "i64": "AwoQOJOvmtzn4H9e3RXl7OzfUhIgOTQzM2Q1MmFlNWY3ZjdmN2U2NzdhZjU0YzllMTcwYTkaDgoFbG9naW4SBWxvZ2lu",
+            "s64": "jAUuaROGZqMCjIoD-BSow9GCMlxELWNjYGArgqeLK0Q"
+        },
+        "bakery-discharge-required": {
+            "m": {
+                "c": [
+                    {
+                        "i": "time-before 2022-03-10T09:26:13.554951585Z"
+                    },
+                    {
+                        "i64": "AwA",
+                        "v64": "84lQ1w2X1y5_pzw6J43_UgGHkExE_T97jn95tTHEZAgaJW3IgRuNI7Yk0qpLM11_kPQ497PxW9yiv6sShpvgUYg7vgPRHi0r",
+                        "l": "https://CONTROLLER_HOST:17070/auth"
+                    }
+                ],
+                "l": "juju model 10c91043-b22b-4e62-8f44-30132106b057",
+                "i64": "AwoQOJOvmtzn4H9e3RXl7OzfUhIgOTQzM2Q1MmFlNWY3ZjdmN2U2NzdhZjU0YzllMTcwYTkaDgoFbG9naW4SBWxvZ2lu",
+                "s64": "jAUuaROGZqMCjIoD-BSow9GCMlxELWNjYGArgqeLK0Q"
+            },
+            "v": 3,
+            "cdata": {
+                "AwA": "A6Ve5aR-tWB-Q0M7ziC2TW_oC9RxW0PwIjQI_eVUhslCKaXcdTmTr5rc5-B_Xt0V5ezs31IttCIUNp99pEQKvgrQDDBRGbJ1SXRGp9WOt8DJpr9dZRfIp_nIG4S0LSzrLnF0TiEy4jZvFcXucApDF7GgPULqnwVX7X_l630n-ZCsTcdZP1W2vnJKsUc_po0jYemzyUBDVckiJz1hDgo"
+            },
+            "ns": "std:"
+        },
+        "discharge-required-error": "invalid login macaroon"
+    }
+}
+```
+
+## 3. Discharging the third party caveat
+
+The client must now try to discharge the caveat, which will prove that the user
+is who they claim to be.  This is done by contacting an auth server (in this
+case implemented on the juju controller) as follows.
+
+This flow is implemented in
+`github.com/go-macaroon-bakery/macaroon-bakery/httpbakery`, in the
+`Client.AcquireDischarge` method.
+
+> Q: Is this the macaroon/Candid flow?
```

Comment:
For candid, we use the ExternalMacaroonAuthenticator but it's the same workflow as far as the client is concerned. The external authenticator hits candid and because it's an external user, ensures the domain in the user tag is "@external" if not set and also does not attempt to look up the user in the juju database cause it won't be there.